File: python/data_structures/stack.py
# data_structures/stack.py
"""
Custom implementation of a Stack (LIFO - Last-In, First-Out).

This implementation uses a Linked List (built from Nodes) as its
underlying structure.

Real-world analogy: A stack of plates. You add (push) a new plate
to the top, and you remove (pop) a plate from the top.
"""
from .node import Node

class Stack:
    """A LIFO Stack implementation using a linked list."""
    
    def __init__(self):
        self.top = None  # Points to the top Node of the stack
        self._size = 0

    def push(self, data):
        """
        Add an item to the top of the stack.
        Complexity: O(1)
        """
        new_node = Node(data)
        new_node.next = self.top
        self.top = new_node
        self._size += 1
        print(f"[Stack]: Pushed {data}")

    def pop(self):
        """
        Remove and return the item from the top of the stack.
        Returns None if the stack is empty.
        Complexity: O(1)
        """
        if self.is_empty():
            return None
        
        data = self.top.data
        self.top = self.top.next  # Move top pointer down
        self._size -= 1
        print(f"[Stack]: Popped {data}")
        return data

    def peek(self):
        """
        Return the item at the top of the stack without removing it.
        Complexity: O(1)
        """
        return self.top.data if self.top else None

    def is_empty(self):
        """Check if the stack is empty. Complexity: O(1)"""
        return self.top is None

    def __len__(self):
        """Return the size of the stack."""
        return self._size

    def __repr__(self):
        items = []
        current = self.top
        while current:
            items.append(str(current.data))
            current = current.next
        return f"Stack(top -> {' -> '.join(items)})"

File: c++/data_structures/Stack.cpp
#pragma once
#include "Node.h"
#include <iostream>
#include <stdexcept> // For std::runtime_error

/*
 * Templated Stack (LIFO) implementation.
 * It is header-only because it's a template.
 * Analogy: A stack of plates.
 */
template <typename T>
class Stack {
private:
    Node<T>* top;
    int _size;

public:
    Stack() : top(nullptr), _size(0) {}

    // Destructor: Essential to prevent memory leaks!
    // It walks the list and deletes every node.
    ~Stack() {
        while (!isEmpty()) {
            pop(); // Pop will delete the node
        }
    }

    // Push an item onto the top of the stack
    // Complexity: O(1)
    void push(T data) {
        Node<T>* newNode = new Node<T>(data);
        newNode->next = top;
        top = newNode;
        _size++;
    }

    // Remove and return the top item
    // Complexity: O(1)
    T pop() {
        if (isEmpty()) {
            throw std::runtime_error("Stack is empty");
        }
        
        Node<T>* temp = top;
        T data = top->data;
        top = top->next;
        
        delete temp; // Free the memory for the node
        _size--;
        
        return data;
    }

    // Return top item without removing
    // Complexity: O(1)
    T peek() const {
        if (isEmpty()) {
            throw std::runtime_error("Stack is empty");
        }
        return top->data;
    }

    bool isEmpty() const {
        return top == nullptr;
    }

    int size() const {
        return _size;
    }
};